In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

X = pd.read_csv("../data/processed/X_train_processed.csv")
X_test = pd.read_csv("../data/processed/X_test_processed.csv")

y = pd.read_csv("../data/processed/y_train_log.csv")
y = y.squeeze()

In [2]:
X_train, X_val, y_train, y_val = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

In [3]:
from sklearn.ensemble import RandomForestRegressor

# Random Forest Regressor
rf_model = RandomForestRegressor(
    n_estimators=200,
    random_state=42,
    n_jobs=-1
)

rf_model.fit(X_train, y_train)

rf_pred = rf_model.predict(X_val)

rf_rmse = np.sqrt(mean_squared_error(y_val, rf_pred))
rf_r2 = r2_score(y_val, rf_pred)

print("Random Forest RMSE log:", rf_rmse)
print("Random Forest R²:", rf_r2)

Random Forest RMSE log: 0.14337599024253278
Random Forest R²: 0.8898421002243425


In [4]:
from sklearn.ensemble import GradientBoostingRegressor

# Gradient Boosting Regressor
gb_model = GradientBoostingRegressor(
    n_estimators=300,
    learning_rate=0.05,
    max_depth=3,
    random_state=42
)

gb_model.fit(X_train, y_train)

gb_pred = gb_model.predict(X_val)

gb_rmse = np.sqrt(mean_squared_error(y_val, gb_pred))
gb_r2 = r2_score(y_val, gb_pred)

print("Gradient Boosting RMSE log:", gb_rmse)
print("Gradient Boosting R²:", gb_r2)

Gradient Boosting RMSE log: 0.13696850252055523
Gradient Boosting R²: 0.8994680279839317


In [5]:
model_results = pd.DataFrame({
    "Model": [
        "Ridge Regression",
        "Random Forest",
        "Gradient Boosting"
    ],
    "RMSE_log": [
        0.121067,
        rf_rmse,
        gb_rmse
    ],
    "R2": [
        0.921455,
        rf_r2,
        gb_r2
    ]
})

model_results.sort_values(by="RMSE_log")

,Model,RMSE_log,R2
0,Ridge Regression,0.121067,0.921455
2,Gradient Boosting,0.136969,0.899468
1,Random Forest,0.143376,0.889842


## Advanced Modeling Summary

In this phase, two advanced models were tested: Random Forest Regressor and Gradient Boosting Regressor.

Random Forest achieved an RMSE_log of 0.143376 and an R² score of 0.889842. Gradient Boosting performed slightly better, with an RMSE_log of 0.136969 and an R² score of 0.899468.

However, neither advanced model outperformed Ridge Regression, which remains the best model so far with the lowest RMSE_log of 0.121067 and the highest R² score of 0.921455.

This suggests that, for the current preprocessing and feature set, a regularized linear model captures the relationship between the features and SalePrice better than the tested tree-based models.